ext/dom/element.c

```diff
@@ -390,31 +393,53 @@ PHP_METHOD(DOMElement, setAttribute)
 
 	DOM_GET_OBJ(nodep, id, xmlNodePtr, intern);
 
-	attr = dom_get_dom1_attribute(nodep, (xmlChar *)name);
-	if (attr != NULL) {
-		switch (attr->type) {
-			case XML_ATTRIBUTE_NODE:
-				node_list_unlink(attr->children);
-				break;
-			case XML_NAMESPACE_DECL:
-				RETURN_FALSE;
-			default:
-				break;
+	if (php_dom_follow_spec_intern(intern)) {
+		xmlChar *name_processed = BAD_CAST name;
+		if (php_dom_ns_is_html_and_document_is_html(nodep)) {
+			char *lowercase_copy = zend_str_tolower_dup_ex(name, name_len);
+			if (lowercase_copy != NULL) {
+				name_processed = BAD_CAST lowercase_copy;
+			}
 		}
 
-	}
+		/* Can't use xmlSetNsProp unconditionally here because that doesn't take into account the qualified name matching... */
+		attr = (xmlNodePtr) php_dom_get_attribute_node(nodep, BAD_CAST name, name_len);
+		if (attr != NULL) {
+			dom_remove_all_children(attr);
+			xmlNodePtr node = xmlNewDocText(attr->doc, BAD_CAST value);
+			xmlAddChild(attr, node);
+		} else {
+			attr = (xmlNodePtr) xmlSetNsProp(nodep, NULL, name_processed, BAD_CAST value);
+		}
 
-	attr = dom_create_attribute(nodep, name, value);
-	if (!attr) {
-		zend_argument_value_error(1, "must be a valid XML attribute");
-		RETURN_THROWS();
-	}
-	if (attr->type == XML_NAMESPACE_DECL) {
-		RETURN_TRUE;
-	}
+		if (name_processed != BAD_CAST name) {
+			efree(name_processed);
+		}
+	} else {
+		attr = dom_get_attribute_or_nsdecl(intern, nodep, BAD_CAST name, name_len);
+		if (attr != NULL) {
+			switch (attr->type) {
+				case XML_ATTRIBUTE_NODE:
+					node_list_unlink(attr->children);
+					break;
+				case XML_NAMESPACE_DECL:
+					RETURN_FALSE;
+				default:
+					break;
```

Comment:
Can this default case happen?

Reply:
No, I believe I just moved this code so I didn't pay attention to that. I can use that one macro though for unreachable default cases.